File: backend/app/api/v1/cert_sync.go
package v1

import (
	"xpanel/app/api/v1/helper"
	"xpanel/app/dto"
	"xpanel/app/service"

	"github.com/gin-gonic/gin"
)

type CertSyncAPI struct{}

// ======================= 证书源管理 =======================

func (a *CertSyncAPI) ListCertSources(c *gin.Context) {
	items, err := service.NewICertSourceService().GetList()
	if err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithData(c, items)
}

func (a *CertSyncAPI) CreateCertSource(c *gin.Context) {
	var req dto.CertSourceCreate
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := service.NewICertSourceService().Create(req); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

func (a *CertSyncAPI) UpdateCertSource(c *gin.Context) {
	var req dto.CertSourceUpdate
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := service.NewICertSourceService().Update(req); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

func (a *CertSyncAPI) DeleteCertSource(c *gin.Context) {
	var req dto.OperateByID
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := service.NewICertSourceService().Delete(req.ID); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

func (a *CertSyncAPI) SyncCertSource(c *gin.Context) {
	var req dto.OperateByID
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := service.NewICertSourceService().Sync(req.ID); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

func (a *CertSyncAPI) TestCertSource(c *gin.Context) {
	var req dto.OperateByID
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := service.NewICertSourceService().TestConnection(req.ID); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

// ======================= 同步日志 =======================

func (a *CertSyncAPI) SearchSyncLogs(c *gin.Context) {
	var req dto.SearchCertSyncLogReq
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	total, items, err := service.NewICertSyncLogService().SearchWithPage(req)
	if err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithPage(c, total, items)
}

// ======================= 证书服务端 =======================

func (a *CertSyncAPI) GetCertServerSetting(c *gin.Context) {
	setting, err := service.NewICertServerService().GetSetting()
	if err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithData(c, setting)
}

func (a *CertSyncAPI) UpdateCertServerSetting(c *gin.Context) {
	var req dto.CertServerSetting
	if err := helper.CheckBindAndValidate(&req, c); err != nil {
		helper.HandleError(c, err)
		return
	}
	if err := service.NewICertServerService().UpdateSetting(req); err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithOutData(c)
}

func (a *CertSyncAPI) ServeCerts(c *gin.Context) {
	items, err := service.NewICertServerService().ListCerts()
	if err != nil {
		helper.HandleError(c, err)
		return
	}
	helper.SuccessWithData(c, items)
}

File: frontend/src/api/modules/cert-sync.ts
import http from '@/api/http'

// --- 证书源管理 ---
export const listCertSources = () => {
  return http.get('/cert-sources')
}

export const createCertSource = (params: {
  name: string
  serverAddr: string
  token: string
  syncInterval: number
  postSyncCommand: string
  enabled: boolean
}) => {
  return http.post('/cert-sources', params)
}

export const updateCertSource = (params: {
  id: number
  name: string
  serverAddr: string
  token?: string
  syncInterval: number
  postSyncCommand: string
  enabled: boolean
}) => {
  return http.post('/cert-sources/update', params)
}

export const deleteCertSource = (id: number) => {
  return http.post('/cert-sources/del', { id })
}

export const syncCertSource = (id: number) => {
  return http.post('/cert-sources/sync', { id })
}

export const testCertSource = (id: number) => {
  return http.post('/cert-sources/test', { id })
}

// --- 同步日志 ---
export const searchSyncLogs = (params: { page: number; pageSize: number; sourceID?: number }) => {
  return http.post('/cert-sync/logs', params)
}

// --- 证书服务端设置 ---
export const getCertServerSetting = () => {
  return http.get('/cert-server/setting')
}

export const updateCertServerSetting = (params: { enabled: boolean; token: string }) => {
  return http.post('/cert-server/setting', params)
}
